Sources/Workspace/ToolsVersionWriter.swift

```diff
@@ -35,6 +35,13 @@ public func writeToolsVersion(at path: AbsolutePath, version: ToolsVersion, fs:
     stream <<< "\n"
     // Append the file contents except for version specifier line.
     stream <<< ToolsVersionLoader.split(contents).rest
+	// FIXME: `split(_:)` above has been deprecated.
+	// However, its replacement (commented out below) causes failures in `FunctionTests` and `WorkspaceTests`.
```

Comment:
If those unit tests aren't testing the right thing or making unwarranted assumptions, then it would be fine to fix them as part of this PR.  But if the tests are arguably correct, then we'd need to figure out how to support the flexibility while not breaking the existing functionality.  I'll take a look at the test failures.

Reply:
I fixed the tests in 06da1ded449207717396c76f73d3322c2995778e. The tests themselves weren't touched, other than a tiny typo fix for a comment.

There were multiple problems with `writeToolsVersion(at:version:fs:)`:

1. It uses the [`<<<` operator](https://github.com/apple/swift-tools-support-core/blob/e129cb07070b7f1be24fc1de413b0989eb0e6744/Sources/TSCBasic/WritableByteStream.swift#L350) to append data in a C++-style. When the right operand is a `Substring`, the compiler selects the [`<<< (stream: WritableByteStream, value: CustomStringConvertible) -> WritableByteStream`](https://github.com/apple/swift-tools-support-core/blob/e129cb07070b7f1be24fc1de413b0989eb0e6744/Sources/TSCBasic/WritableByteStream.swift#L372) overload as the callee. This overload appends the `description` of the operand, so the string `"0 bytes"` is appended whenever the operand is empty. This is fixed by creating a `ByteString` from the `Substring`, so the [correct overload](https://github.com/apple/swift-tools-support-core/blob/e129cb07070b7f1be24fc1de413b0989eb0e6744/Sources/TSCBasic/WritableByteStream.swift#L378) is selected.

2. The old (now removed) `ToolsVersionLoader.split(_:)` [returns the entire manifest](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/49cfc46bc5defd3ce8e0c0261e3e2cb475bcdb91/Sources/PackageLoading/ToolsVersionLoader.swift#L170) as the contents after the Swift tools version specification line, whenever the line is invalid according to the old rule. On the contrary, the new `ToolsVersionLoader.split(_:)` returns the actual contents after the Swift tools version specification line as the contents after the Swift tools version specification line, regardless of the specification's validity or backward-compatibility. This difference is resolved with [a control flow](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/06da1ded449207717396c76f73d3322c2995778e/Sources/Workspace/ToolsVersionWriter.swift#L49) conditional over [`ToolsVersionLoader.ManifestComponents.toolsVersionSpecificationComponents.everythingUpToVersionSpecifierIsWellFormed`](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/06da1ded449207717396c76f73d3322c2995778e/Sources/PackageLoading/ToolsVersionLoader.swift#L282).
   
   `everythingUpToVersionSpecifierIsWellFormed` is sufficient because the old `split(_:)` treats the version specifier as a "anything goes" field ([`.*?` in `regex`](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/49cfc46bc5defd3ce8e0c0261e3e2cb475bcdb91/Sources/PackageLoading/ToolsVersionLoader.swift#L182)), only caring about everything before the version specifier.

3. Because the version specifier is "anything goes" in the old `split(_:)`, if a Swift tools version specification is like this:
   
   ```swift
   // swift-tools-version:swift 5.0
   ```
   
   or this:
   
   ```swift
   // swift-tools-version:::::5.0
   ```
   
   it would be considered of having a valid label `"swift-tools-version:"` followed by `"swift 5.0"` or `"::::5.0"` as the version specifier. 
   
   The new split(_:), on the other hand, would consider that the label `"swift-tools-version:swift"` or `"swift-tools-version:::::"` is invalid, followed by `"5.0"` as the version specifier.
   
   This difference in behaviour is resolved with a slightly altered logic in the new `split(_:)` that introduces a special handling for when the label is prefixed with `"swift-tools-versioin:"` (case-insensitive). This change sort of removed some ambiguity in the diagnosis of misspelt labels, but introduced more ambiguity in the diagnosis of missing labels. It's hard for me to say whether this change is for the better or the worse, other than it made `ToolsVersionWriterTests` work with the new `split(_:)`.
   
   There is some more detailed documentation in the source [here](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/06da1ded449207717396c76f73d3322c2995778e/Sources/PackageLoading/ToolsVersionLoader.swift#L511) and [here](https://github.com/WowbaggersLiquidLunch/swift-package-manager/blob/06da1ded449207717396c76f73d3322c2995778e/Sources/PackageLoading/ToolsVersionLoader.swift#L128).

Reply:
Thanks for the detailed consideration of this.  I think that as long as the error clearly points out that there is a problem with the tools version, which it does, then in the case of ambiguity I think it's fine that it might misunderstand the specific nature of the syntax error.  The syntax is fairly clear and by comparing with a valid example, I think people will be able to understand what they need to fix.